services/httpd/msgp_emitter.go

```diff
@@ -0,0 +1,160 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type messagePackEncoder struct {
+	Epoch     string
+	ChunkSize int
+}
+
+func (e *messagePackEncoder) ContentType() string {
+	return "application/x-msgpack"
+}
+
+func (e *messagePackEncoder) Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet) {
+	var convertToEpoch func(row *influxql.Row)
+	if e.Epoch != "" {
+		convertToEpoch = epochConverter(e.Epoch)
+	}
+	values := make([][]interface{}, 0, e.ChunkSize)
+
+	enc := msgp.NewWriter(w)
+	enc.WriteMapHeader(1)
+	enc.WriteString("results")
+	enc.WriteInt(header.Results)
+
+	enc.Flush()
+	if w, ok := w.(http.Flusher); ok {
+		w.Flush()
```

Comment:
Why do we need to flush after a certain type of data is sent, instead of flushing when buffer is full?
Or alternatively, what is the purpose of flushing headers when there is no data? The client can read the headers yes, but if it has no data, the headers don't serve much purpose.